core/arch/arm/plat-stm32mp1/nsec-service/stm32mp1_svc_setup.c

```diff
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (c) 2017-2018, STMicroelectronics
+ */
+
+#include <arm.h>
+#include <drivers/stm32_bsec.h>
+#include <sm/optee_smc.h>
+#include <sm/sm.h>
+#include <string.h>
+#include <stm32_util.h>
+
+#include "bsec_svc.h"
+#include "stm32mp1_smc.h"
+
+/* STM32 SiP Service UUID */
+static const TEE_UUID stm32mp1_sip_svc_uid = {
+	0x50aa78a7, 0x9bf4, 0x4a14,
+	{ 0x8a, 0x5e, 0x26, 0x4d, 0x59, 0x94, 0xc2, 0x14 }
+};
+
+static void get_sip_func_uid(uint32_t *a0, uint32_t *a1,
+			     uint32_t *a2, uint32_t *a3)
+{
+	const void *uid = &stm32mp1_sip_svc_uid;
+
+	memcpy(a0, (char *)uid, sizeof(uint32_t));
+	memcpy(a1, (char *)uid + sizeof(uint32_t), sizeof(uint32_t));
+	memcpy(a2, (char *)uid + (sizeof(uint32_t) * 2), sizeof(uint32_t));
+	memcpy(a3, (char *)uid + (sizeof(uint32_t) * 3), sizeof(uint32_t));
+}
+
+static bool sip_service(struct sm_ctx *ctx __unused,
+			uint32_t *a0, uint32_t *a1, uint32_t *a2, uint32_t *a3)
+{
+	switch (OPTEE_SMC_FUNC_NUM(*a0)) {
+	case STM32_SIP_FUNC_VERSION:
+		*a0 = STM32_SIP_SVC_VERSION_MAJOR;
+		*a1 = STM32_SIP_SVC_VERSION_MINOR;
+		break;
+	case STM32_SIP_FUNC_UID:
+		get_sip_func_uid(a0, a1, a2, a3);
+		break;
+#ifdef CFG_STM32_BSEC_SIP
+	case STM32_SIP_FUNC_BSEC:
+		*a0 = bsec_main(*a1, *a2, *a3, a1);
+		break;
+#endif
+	default:
+		return true;
+	}
+
+	return false;
+}
+
+static bool oem_service(struct sm_ctx *ctx __unused,
+			uint32_t *a0 __unused, uint32_t *a1 __unused,
+			uint32_t *a2 __unused, uint32_t *a3 __unused)
+{
+	return true;
+}
+
+/* Override default sm_platform_handler() with paltform specific function */
+bool sm_platform_handler(struct sm_ctx *ctx)
+{
+	uint32_t *a0 = (uint32_t *)(&ctx->nsec.r0);
+	uint32_t *a1 = (uint32_t *)(&ctx->nsec.r1);
+	uint32_t *a2 = (uint32_t *)(&ctx->nsec.r2);
+	uint32_t *a3 = (uint32_t *)(&ctx->nsec.r3);
+
+	if (!OPTEE_SMC_IS_FAST_CALL(*a0))
+		return true;
```

Comment:
A comment explaining what happens when `true` or `false` is returned would be nice. The strategy here seems to let everything unknown (even if it would fall in the SIP or OEM range) through to `tee_entry_fast()` which then will return `OPTEE_SMC_RETURN_UNKNOWN_FUNCTION` in `a0`.

Reply:
Ok, I will add inline  comments to describes that.
(I'll create later a P-R to have the generic `sm_platform_handler()` return value be more consistent: return true when SMC was handled, false when not).

Reply:
The return value means "switch to secure mode", this happens to be the inverse of "SMC was handled" so far.
We should perhaps have an enum instead to make it more clear, for instance something like:
``` C
enum sm_handler_ret {
    SM_HANDLER_SMC_HANDLED  = 0,
    SM_HANDLER_SWITCH_TO_SEC_WORLD,
};
```

Reply:
+1, i'll do that.

Reply:
See P-R https://github.com/OP-TEE/optee_os/pull/3069